sympy/sets/fancysets.py

```diff
@@ -576,50 +598,62 @@ def reversed(self):
 
     def _contains(self, other):
+        from sympy.functions.elementary.piecewise import Piecewise
+        from sympy.functions.elementary.integers import ceiling
+        other = _sympify(other)
         if not self:
             return S.false
         if other.is_infinite:
             return S.false
         if not other.is_integer:
             return other.is_integer
         ref = self.start if self.start.is_finite else self.stop
-        if (ref - other) % self.step:  # off sequence
-            return S.false
-        return _sympify(other >= self.inf and other <= self.sup)
+        start, stop, step = self.start, self.stop, self.step
+        inf = Piecewise((start, Ge(step, S.Zero)), (stop - step, True))
+        sup = Piecewise((stop - step, Ge(step, S.Zero)), (start, True))
+        return Piecewise((False, Le(ceiling((stop - start)/step), S.Zero)),
+                         (False, Ne((ref - other) % self.step, S.Zero)),
+                         (True, Ge(other, inf) & Le(other, sup)),
+                         (False, True))
 
     def __iter__(self):
         if self.start in [S.NegativeInfinity, S.Infinity]:
             raise ValueError("Cannot iterate over Range with infinite start")
-        elif self:
+        if (self.size.is_Integer != True) and (self.size not in [S.Infinity, S.NegativeInfinity]):
+            raise ValueError("Cannot iterate over symbolic sized range.")
+        if self:
             i = self.start
             step = self.step
 
-            while True:
-                if (step > 0 and not (self.start <= i < self.stop)) or \
-                   (step < 0 and not (self.stop < i <= self.start)):
-                    break
+            while And(Or(Le(step, 0), And(Le(self.start, i), Lt(i, self.stop))),
+                     Or(Ge(step, 0), And(Lt(self.stop, i), Le(i, self.start)))) != False:
                 yield i
                 i += step
 
     def __len__(self):
-        if not self:
-            return 0
-        dif = self.stop - self.start
-        if dif.is_infinite:
+        size = _sympify(self.size)
+        if size.has(Symbol) or size.is_infinite:
             raise ValueError(
-                "Use .size to get the length of an infinite Range")
-        return abs(dif//self.step)
+                "Use .size to get the length of an infinite and symbolic Range")
+        return int(size)
 
     @property
     def size(self):
-        try:
-            return _sympify(len(self))
-        except ValueError:
+        from sympy.functions.elementary.integers import floor, ceiling
+        from sympy.functions.elementary.piecewise import Piecewise
+        if not self:
+            return S.Zero
+        start, stop, step = self.start, self.stop, self.step
+        dif = stop - start
+        if dif.is_infinite:
             return S.Infinity
+        null = ceiling(dif/step)
+        return Piecewise((abs(floor(dif/step)), Gt(null, S.Zero)),
+                         (S.Zero, True))
```

Comment:
Does this Piecewise also handles the S.Zero and S.Infinity cases above so that they aren't needed any more?